CONTRIBUTING.md

```diff
@@ -66,6 +66,111 @@ For first-time contributers, we suggest to look for issues categorized as [trivi
 
 If you want a hassle-free contribution look for issues categorized as [preapproved](https://issues.dlang.org/buglist.cgi?component=dmd&keywords=preapproved&product=D).
 
+## DMD Style
+
+Here is a short list of stylistic issues the core team will expect in
+pull requests. Much of the source code does not follow these, but
+we expect new code to, and PRs retrofitting the existing code to
+follow it is welcome.
+
+1. Use attributes `const`/`nothrow`/`pure`/`scope`/`@safe`/`private`/etc.
+Successfully using `pure` functions is regarded with particular favor.
+
+2. Use correct Ddoc function comment blocks. Do not use Ddoc comments for
+overrides unless the overriding function does something different (as far as
+the caller is concerned) than the overridden function. Ddoc comment blocks
+are often overkill for nested functions and function literals; use ordinary
+comments for those.
+
+3. Do not use `strlen`/`strcmp` and their ilk. Use D arrays instead.
+
+4. Use `ref` instead of raw pointers.
+
+5. Use nested functions to get rid of rats' nests of goto's.
+
+6. Look for duplicative code and factor out into functions.
+
+7. Declare local variables as `const` as much as possible.
+
+8. Use Single Assignment for local variables:
+```
+T t = x;
+...
+t = y;
+ ...
+```
+becomes:
+```
+T tx = x;
+...
+T ty = y;
+...
+```
+
+9. "Shrinkwrap" the scope of local variables as tightly as possible
+around their uses.
+
+10. Similar to (8), use distinct variable names for non-overlapping uses.
+
+11. Avoid use of mutable globals as much as practical. Consider passing them
+in as parameters.
+
+12. Avoid use of default parameters. Spell them out.
+
+13. Minimize use of overloading.
+
+14. Avoid clever code. Anybody can write clever code. It takes genius to write
+simple code.
+
+15. Try to reduce cyclomatic complexity, i.e. think about how to make the code work
+without control flow statements.
+
+16. Try not to mix functions that "answer a question" with functions that
+"mutate the data".
+This was done successfully in src/dmd/escape.d, it wasn't easy, but
+it was well worth it.
+
+17. Try to eliminate reliance on `global.errors`.
+
+18. For aggregates that expose public access to fields, think hard about why this is
+necessary and if it can be done better. Merely replacing them with read/write properties
+accomplishes nothing. The more of its internal details can be made private, the better.
+
+19. Try to use function prefixes:
+
+ * `is` is the parameter in a certain category?
+ * `has` does the parameter have a certain feature?
+ * `can` for can I do X with the parameter?
+
+Such functions should not be mutating the data nor issuing error messages.
+
+20. The function return value variable should be named `result`.
+
+21. The more constrained the scope of a name is, the shorter it should be.
+
+
+
+## The following will not be viewed with favor:
+
+1. Grand renamings of variables and functions
```

Comment:
I'll put it another way. Changes that go against the recommendations here need my approval.

Reply:
We have [The D Style](https://dlang.org/dstyle.html), and DMD doesn't follow it.  That's a problem, IMO.  Why can't we do something about it and bring Phobos, druntime and DMD all under the same style?  You are always welcome to make a PR against The D Style document to improve it if there's something you don't like about it.  Furthermore, if DMD were updated to be consistent with The D Style, we could enable DScanner linting checks to enforce that it stays that way now and into the future.

Reply:
I don't have any problem with the D Style. I'm afraid, though, of opening the floodgates to massive PRs to re-style the code, which is an **Illusion of Progress**. It won't remove one iota of technical debt, but it'll sure be disruptive and waste everyone's time.

Reply:
Yes, it will open up the floodgates to massive PRs **temporarily**.  Let's get it over and done with.

Reply:
And I don't agree that there will be no benefit.  After the code is converted to The D Style, we can enable DScanner linter checks to enforce that it remains that way.  @andralex has repeatedly complained that our reviews often do not contain any substance and instead are distracted by discussions about formatting, style, and conventions.  Once this work is done, the DScanner linter will take care of a lot of those issues.  It will remove the formatting, styling, and convention distractions so we can have better code reviews.  Yes, sometimes it will be disruptive, but I don't agree that it will be **very** disruptive, and it will only last until the work is done.

Reply:
There have been multiple waves of these types of changes in both dmd and Phobos - like tax increases, it never is over and done with. There's always the next styling fashion trivia.

The DMD source code is **not** poorly styled (it's just old fashioned) and the styling is not the source of its technical debt nor bugs.

This long thread is **exactly** why I don't want to put this requirement in this PR. It's just a fat waste of time, and putting it in will spawn many more of these threads. I had a long conversation with Andrei today about the difficulties I was having in const-ifying the dmd source code and ways to deal with that. That's the kind of thing we should be talking about, rather than should we paint the room with polka dots or stripes.

Reply:
Adding `const` doesn't fix any bugs or adds any new features. It could be considered what you call Illusion of Progress as well.

Reply:
@jacob-carlborg, on the contrary, I'd argue that as long as `const` is not casted away (thereby eroding its meaning), `const` helps bring clarity in dmd's codebase particularly because when you call function you have no idea whether it may may modify some completely unrelated object that you have reference to.

Reply:
> There have been multiple waves of these types of changes in both dmd and Phobos - like tax increases, it never is over and done with. There's always the next .

@WalterBright As far as I have observed, https://dlang.org/dstyle has been primarily driven by @andralex or changes to it have been otherwise approved by him. I haven't seen any back and forth changes to that page, and for the most part all changes have been filling the gaps. Most of DStyle changes were simply codifying existing practices with very few exceptions (IIRC mostly proposed by @andralex) like not indenting `if` constraints, space after `cast(T)`, etc. 
For Phobos, most of the changes were done in several pull requests in 2017 and most of the work was not changing the existing code, but setting up the continuous integration to enforce it. Since at least 6-7 months this is a closed case.
The changes were not disruptive at all, on the contrary - they helped save time as before there were many pull requests by newcomers (specifically for `core/sys` and `core/stdc` in druntime) that followed the style of whatever the nearby code was, but back then even in a single file you could find several different styles. After we fixed that in a small number of large PRs we haven't had issues with this anymore.

Reply:
At Facebook grand renames were rare, discouraged, and required approval by the tech leads. They used special tooling too so as to avoid false blame. Interns or starting engineers would regularly find reasons to propose them and they'd be brought back in line.

That said, they did happen. I recall the last I've done replaced `NULL` with `nullptr_t`. It found a bunch of cases in which `NULL` was being assigned to an integral (lol) and introduced a difficult to find bug (I forgot the particulars, might have been a function call).

There are many, many things to improve in dmd's codebase. Grand renamings is among the most attractive and of the lowest impact. I am shocked that anyone would consider making more things immutable (immutable is better than const) is considered low impact.

Reply:
>And I don't agree that there will be no benefit. After the code is converted to The D Style, we can enable DScanner linter checks to enforce that it remains that way.

I think this is worthwhile. Do we have the linting infra ready to go?

Reply:
> I am shocked that anyone would consider making more things immutable (immutable is better than const) is considered low impact.

I tried to make changes that allowed to add `pure`, which will improve the code base, **and** adding more features. That got shut down immediately. Adding `const` or `immutable` by it self will not add new features or fix bugs.

Reply:
> I think this is worthwhile. Do we have the linting infra ready to go?

For a long time already: https://github.com/dlang/dmd/blob/master/src/.dscanner.ini (most tests are just disabled)

We also stopped to enforce more style checks at Phobos as everyone lost interest in this, but there's still a lot that could be done:

https://github.com/dlang/phobos/blob/master/.dscanner.ini (disabled checks are at the bottom)

Reply:
> Do we have the linting infra ready to go?

https://github.com/dlang/dmd/blob/874f5da650b73025db793e7ac9a2c1ee4e0fedad/src/posix.mak#L572-L594

Reply:
@jacob-carlborg maybe it's the "and" that's the problem?

Reply:
@wilzbach cool, that stuff should be at least enabled for new PRs.

Reply:
> maybe it's the "and" that's the problem?

I don't know you tell me. It's the "and" that makes it useful. It was not several changes that were combined in a single PR. It was preparation work that would allow to add `pure` and then add more features as well.

Reply:
@jacob-carlborg I recall @WalterBright rejected some work from you that was too large for its usefulness. Was that it? I think he had valid reasons. At any rate, if some work you did was rejected that doesn't compromise the notion that e.g. more pure is better.

Reply:
It's impossible to make any useful changes in the DMD code base without large refactoring. The things you are suggesting and Walter is doing, i.e. adding `const` in one or two places is not very useful. It doesn't allow the compiler to do anything that wasn't possible before.

Reply:
>It's impossible to make any useful changes in the DMD code base without large refactoring.

I disagree with this.

>The things you are suggesting and Walter is doing, i.e. adding const in one or two places is not very useful.

This, too. Hard disagree.

Reply:
> @wilzbach cool, that stuff should be at least enabled for new PRs.

No, can't do. It would then lead to new PRs needing to make those unwanted "unrelated changes".
It needs to be enabled separately (either for the entire project or specific modules).

Reply:
> I disagree with this.

We have very different view of that "useful" is. I'm going to stop replying now since this is completely pointless and doesn't lead anywhere.

Reply:
>We have very different view of that "useful" is.

That creates opportunity.

Reply:
@wilzbach I understand. Style adjustments for uniformity have usefulness because they eliminate certain debate going forward, and are enforceable automatically. Please send such my way (starting  with the most useful/least disruptive) and I'll have them in. Thanks.

Reply:
Sorry, I gave up on these. I have more important fights to fight these days.